### PR TITLE
MusicXML Parsing: Allow implicit measure numbers to not show up.

### PR DESCRIPTION
**Pull request by earboxer**
Goal:

musicXML
```xml
<measure number="1" implicit="yes">
```

MEI
```xml
<measure n="1">
  <mNum></mNum>
```

See https://github.com/rism-ch/verovio/issues/1153#issuecomment-540150400

### CODE REVIEW

**Comment by craigsapp** [comment]
> Screen Shot 2019-10-10 at 8 36 58 AM

Is there a point to this particular goal?  In other words, "measure 1" is always implicit in musical scores that have measure numbers. 

In verovio, measure 1 is also always "implicit" (not shown on the score).  In the following example I added `<mNum>X</mNum>` to the first measure:

Screen Shot 2019-10-10 at 9 09 05 AM

Probably the best thing would be for verovio to change its behavior: When `<mNum>` is present, this number/text will be forced to be displayed.  Currently this is true for all measures except measure `@n=1` (and `@n=0`).  So display of `<mNum>` on those measures should be added (or more precisely, not suppressed).   Here is what the notation looks like when I set the first measure to `@n="2"` rather than `@n="1"`:

Screen Shot 2019-10-10 at 9 21 12 AM

Is there a way to suppress display of `<mNum>` measure numbers?   At some point someone will want such a feature: this would allow encoding of the originally visible measure numbers in a manuscript/print score, but allow a different measure number display style from the encoded score without deleting the original information.  Currently using `<mNum>` with verovio will result in problems, except for the case of suppressing measure numbers.  When measure numbers should only be shown at the start of the system, then the visibility of the measure number is dependent on the layout, which is not determinable beforehand unless explicit system breaks are also encoded.

Should verovio's behavior change, such that the presence of any `<mNum>` in any measure causes suppression of any automatically displayed barlines (such as at the start of each system)?  This would minimize conflict between the automatic/explicit measure number display.  And conversely, there should be an option for the reverse case: suppress `<mNum>` when automatically displaying measure numbers is desired.

Also to consider: when MusicXML is encoding explicit/implicit measure numbers, what does this mean? It is explicitly indicating some automated system?  In that case it is better to reconstruct the automated system.  For example, what happens when you add a measure into the middle of a work in Finale/Sibelius/Sibelius?  I would think they all would automatically renumber the measures, and any system for displaying barlines in different styles would automatically update as well.












Test MEI data:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="http://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="http://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="4.0.0">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2019-10-10T08:56:59" version="2.2.0-dev-97d3572-dirty">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="mdiv-0000001229074009">
    <score xml:id="score-0000000401668770">
     <scoreDef xml:id="scoredef-0000001888816109" midi.bpm="400">
      <staffGrp xml:id="staffgrp-0000001713290013">
       <staffDef xml:id="staffdef-0000001228330943" n="1" lines="5">
        <clef xml:id="clef-0000000279161489" shape="G" line="2" />
        <meterSig xml:id="metersig-L2F1" count="4" unit="4" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1" n="1">
      <mNum>X</mNum>
       <staff xml:id="staff-0000001798449276" n="1"> 
        <layer xml:id="layer-L1F1N1" n="1">
         <note xml:id="note-L4F1" dur="1" oct="4" pname="c" accid.ges="n" />
        </layer>
       </staff>
      </measure>
      <measure xml:id="measure-L5" right="end" n="2">
       <mNum>Y</mNum>
       <staff xml:id="staff-L5F1N1" n="1">
        <layer xml:id="layer-L5F1N1" n="1">
         <note xml:id="note-L6F1" dur="1" oct="4" pname="d" accid.ges="n" />
        </layer>
       </staff>
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```


Test data for second example:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="http://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="http://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="4.0.0">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2019-10-10T09:20:16" version="2.2.0-dev-97d3572-dirty">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="mdiv-0000001821415851">
    <score xml:id="score-0000000156819772">
     <scoreDef xml:id="scoredef-0000001318088608" midi.bpm="400">
      <staffGrp xml:id="staffgrp-0000001038774327">
       <staffDef xml:id="staffdef-0000002031510153" n="1" lines="5">
        <clef xml:id="clef-0000002122966293" shape="G" line="2" />
        <meterSig xml:id="metersig-L2F1" count="4" unit="4" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1" n="2">
       <mNum>Y</mNum>
       <staff xml:id="staff-0000002002773842" n="1">
        <layer xml:id="layer-L1F1N1" n="1">
         <note xml:id="note-L4F1" dur="1" oct="4" pname="c" accid.ges="n" />
        </layer>
       </staff>
      </measure>
      <measure xml:id="measure-L5" right="end" n="3">
       <mNum>Z</mNum>
       <staff xml:id="staff-L5F1N1" n="1">
        <layer xml:id="layer-L5F1N1" n="1">
         <note xml:id="note-L6F1" dur="1" oct="4" pname="d" accid.ges="n" />
        </layer>
       </staff>
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

**Comment by earboxer** [comment]
> Probably the best thing would be for verovio to change its behavior: ... So display of <mNum> on those measures should be added (or more precisely, not suppressed)

I agree. (Though it should not happen before this PR is merged)

> Is there a way to suppress display of <mNum> measure numbers? At some point someone will want such a feature: this would allow encoding of the originally visible measure numbers in a manuscript/print score, but allow a different measure number display style from the encoded score without deleting the original information.

I think this is out of scope: can be fixed by changing your MEI input (XSLT stylesheet or other means).


> Should verovio's behavior change, such that the presence of any <mNum> in any measure causes suppression of any automatically displayed barlines (such as at the start of each system)

I think this is a little too complicated, and not what people would expect based on the MEI spec:

>> (measure number) – Designation, name, or label for a measure, often but not always consisting of digits. Use this element when the n attribute on measure does not adequately capture the appearance or placement of the measure number/label. 

https://music-encoding.org/guidelines/v4/elements/mnum.html

**Comment by earboxer** [comment]
> Also to consider: when MusicXML is encoding explicit/implicit measure numbers, what does this mean? It is explicitly indicating some automated system? In that case it is better to reconstruct the automated system. For example, what happens when you add a measure into the middle of a work in Finale/Sibelius/Sibelius? I would think they all would automatically renumber the measures, and any system for displaying barlines in different styles would automatically update as well.

**MusicXML** can't represent explicit measure numbers (i.e. the opposite of `implicit="yes"`)(like #909 wants, or as seen in https://github.com/rism-ch/verovio/issues/1153#issuecomment-540050375).  In MuseScore, if you insert a measure into a score, you can check "Exclude from measure count" under measure properties, which will then export musicXML with `<measure number="X1" implicit="yes">`. (Which works fine in Verovio)

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE score-partwise PUBLIC "-//Recordare//DTD MusicXML 3.1 Partwise//EN" "http://www.musicxml.org/dtds/partwise.dtd">
<score-partwise version="3.1">
  <work>
    <work-title>Title</work-title>
    </work>
  <identification>
    <creator type="composer">Composer</creator>
    <encoding>
      <software>MuseScore 3.2.3</software>
      <encoding-date>2019-10-10</encoding-date>
      <supports element="accidental" type="yes"/>
      <supports element="beam" type="yes"/>
      <supports element="print" attribute="new-page" type="yes" value="yes"/>
      <supports element="print" attribute="new-system" type="yes" value="yes"/>
      <supports element="stem" type="yes"/>
      </encoding>
    </identification>
  <defaults>
    <scaling>
      <millimeters>7.05556</millimeters>
      <tenths>40</tenths>
      </scaling>
    <page-layout>
      <page-height>1683.78</page-height>
      <page-width>1190.55</page-width>
      <page-margins type="even">
        <left-margin>56.6929</left-margin>
        <right-margin>56.6929</right-margin>
        <top-margin>56.6929</top-margin>
        <bottom-margin>113.386</bottom-margin>
        </page-margins>
      <page-margins type="odd">
        <left-margin>56.6929</left-margin>
        <right-margin>56.6929</right-margin>
        <top-margin>56.6929</top-margin>
        <bottom-margin>113.386</bottom-margin>
        </page-margins>
      </page-layout>
    <word-font font-family="FreeSerif" font-size="10"/>
    <lyric-font font-family="FreeSerif" font-size="11"/>
    </defaults>
  <credit page="1">
    <credit-words default-x="1133.86" default-y="1502.09" justify="right" valign="bottom" font-size="12">Composer</credit-words>
    </credit>
  <credit page="1">
    <credit-words default-x="595.275" default-y="1627.09" justify="center" valign="top" font-size="24">Title</credit-words>
    </credit>
  <part-list>
    <score-part id="P1">
      <part-name>Piano</part-name>
      <part-abbreviation>Pno.</part-abbreviation>
      <score-instrument id="P1-I1">
        <instrument-name>Piano</instrument-name>
        </score-instrument>
      <midi-device id="P1-I1" port="1"></midi-device>
      <midi-instrument id="P1-I1">
        <midi-channel>1</midi-channel>
        <midi-program>1</midi-program>
        <volume>78.7402</volume>
        <pan>0</pan>
        </midi-instrument>
      </score-part>
    </part-list>
  <part id="P1">
    <measure number="1" width="276.51">
      <print>
        <system-layout>
          <system-margins>
            <left-margin>0.00</left-margin>
            <right-margin>0.00</right-margin>
            </system-margins>
          <top-system-distance>195.00</top-system-distance>
          </system-layout>
        </print>
      <attributes>
        <divisions>1</divisions>
        <key>
          <fifths>0</fifths>
          </key>
        <time>
          <beats>4</beats>
          <beat-type>4</beat-type>
          </time>
        <clef>
          <sign>G</sign>
          <line>2</line>
          </clef>
        </attributes>
      <note default-x="69.27" default-y="-25.00">
        <pitch>
          <step>A</step>
          <octave>4</octave>
          </pitch>
        <duration>4</duration>
        <voice>1</voice>
        <type>whole</type>
        <lyric number="1" default-x="3.26" default-y="-40.62" relative-y="-30.00">
          <syllabic>single</syllabic>
          <text>1</text>
          </lyric>
        </note>
      </measure>
    <measure number="2" width="221.86">
      <note default-x="13.32" default-y="-20.00">
        <pitch>
          <step>B</step>
          <octave>4</octave>
          </pitch>
        <duration>4</duration>
        <voice>1</voice>
        <type>whole</type>
        <lyric number="1" default-x="9.90" default-y="-40.62" relative-y="-30.00">
          <syllabic>single</syllabic>
          <text>2</text>
          </lyric>
        </note>
      </measure>
    <measure number="X1" implicit="yes" width="353.21">
      <note default-x="18.79" default-y="-40.00">
        <pitch>
          <step>E</step>
          <octave>4</octave>
          </pitch>
        <duration>1</duration>
        <voice>1</voice>
        <type>quarter</type>
        <stem>up</stem>
        <lyric number="1" default-x="6.58" default-y="-40.62" relative-y="-30.00">
          <syllabic>begin</syllabic>
          <text>add</text>
          </lyric>
        </note>
      <note default-x="102.00" default-y="-40.00">
        <pitch>
          <step>E</step>
          <octave>4</octave>
          </pitch>
        <duration>1</duration>
        <voice>1</voice>
        <type>quarter</type>
        <stem>up</stem>
        <lyric number="1" default-x="6.58" default-y="-40.62" relative-y="-30.00">
          <syllabic>end</syllabic>
          <text>ed</text>
          </lyric>
        </note>
      <note default-x="185.20" default-y="-40.00">
        <pitch>
          <step>E</step>
          <octave>4</octave>
          </pitch>
        <duration>1</duration>
        <voice>1</voice>
        <type>quarter</type>
        <stem>up</stem>
        <lyric number="1" default-x="6.58" default-y="-40.62" relative-y="-30.00">
          <syllabic>begin</syllabic>
          <text>mea</text>
          </lyric>
        </note>
      <note default-x="268.41" default-y="-40.00">
        <pitch>
          <step>E</step>
          <octave>4</octave>
          </pitch>
        <duration>1</duration>
        <voice>1</voice>
        <type>quarter</type>
        <stem>up</stem>
        <lyric number="1" default-x="6.58" default-y="-40.62" relative-y="-30.00">
          <syllabic>end</syllabic>
          <text>sure</text>
          </lyric>
        </note>
      </measure>
    <measure number="3" width="225.58">
      <note default-x="13.32" default-y="-15.00">
        <pitch>
          <step>C</step>
          <octave>5</octave>
          </pitch>
        <duration>4</duration>
        <voice>1</voice>
        <type>whole</type>
        <lyric number="1" default-x="9.90" default-y="-40.62" relative-y="-30.00">
          <syllabic>single</syllabic>
          <text>3</text>
          </lyric>
        </note>
      <barline location="right">
        <bar-style>light-heavy</bar-style>
        </barline>
      </measure>
    </part>
  </score-partwise>
```

Verovio has no problem with this, but produces technically invalid MEI
```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="http://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="http://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="4.0.0">
    <meiHead>
        <fileDesc>
            <titleStmt>
                <title>Title</title>
                <respStmt>
                    <persName role="composer">Composer</persName>
                </respStmt>
            </titleStmt>
            <pubStmt></pubStmt>
        </fileDesc>
        <encodingDesc xml:id="encodingdesc-0000000251296704">
            <appInfo xml:id="appinfo-0000001590854126">
                <application xml:id="application-0000001313890532" isodate="2019-10-10T14:03:24" version="2.2.0-dev-6e1ea6c-dirty">
                    <name xml:id="name-0000002131312870">Verovio</name>
                    <p xml:id="p-0000000948174130">Transcoded from MusicXML</p>
                </application>
            </appInfo>
        </encodingDesc>
    </meiHead>
    <music>
        <body>
            <mdiv xml:id="mdiv-0000001633942170">
                <score xml:id="score-0000001792657001">
                    <scoreDef xml:id="scoredef-0000000252788857">
                        <staffGrp xml:id="staffgrp-0000001773364083">
                            <staffDef xml:id="staffdef-0000001488827048" n="1" lines="5" ppq="1" clef.shape="G" clef.line="2" meter.count="4" meter.unit="4">
                                <keySig xml:id="keysig-0000000028631783" sig="0" />
                                <label xml:id="label-0000000004606268">Piano</label>
                                <labelAbbr xml:id="labelAbbr-0000000653510726">Pno.</labelAbbr>
                                <instrDef xml:id="instrdef-0000000914559089" midi.channel="0" midi.instrnum="0" midi.volume="78.00%" />
                            </staffDef>
                        </staffGrp>
                    </scoreDef>
                    <section xml:id="section-0000002138132044">
                        <pb xml:id="pb-0000001741398257" />
                        <measure xml:id="measure-0000000872769800" n="1">
                            <staff xml:id="staff-0000000564227364" n="1">
                                <layer xml:id="layer-0000001829005243" n="1">
                                    <note xml:id="note-0000001010195943" dur.ppq="4" dur="1" oct="4" pname="a">
                                        <verse xml:id="verse-0000000357500819" n="1">
                                            <syl xml:id="syl-0000002004504274" con="s">1</syl>
                                        </verse>
                                    </note>
                                </layer>
                            </staff>
                        </measure>
                        <measure xml:id="measure-0000001128466700" n="2">
                            <staff xml:id="staff-0000000894002769" n="1">
                                <layer xml:id="layer-0000001708944171" n="1">
                                    <note xml:id="note-0000001778387019" dur.ppq="4" dur="1" oct="4" pname="b">
                                        <verse xml:id="verse-0000000673229387" n="1">
                                            <syl xml:id="syl-0000002022454913" con="s">2</syl>
                                        </verse>
                                    </note>
                                </layer>
                            </staff>
                        </measure>
                        <measure xml:id="measure-0000001879691822" n="X1">
                            <mNum xml:id="mnum-0000001642981700" />
                            <staff xml:id="staff-0000001248698774" n="1">
                                <layer xml:id="layer-0000001670096134" n="1">
                                    <note xml:id="note-0000001694457848" dur.ppq="1" dur="4" oct="4" pname="e" stem.dir="up">
                                        <verse xml:id="verse-0000000972408469" n="1">
                                            <syl xml:id="syl-0000000918584813" con="d" wordpos="i">add</syl>
                                        </verse>
                                    </note>
                                    <note xml:id="note-0000001125118179" dur.ppq="1" dur="4" oct="4" pname="e" stem.dir="up">
                                        <verse xml:id="verse-0000001267722618" n="1">
                                            <syl xml:id="syl-0000001428778839" con="s" wordpos="t">ed</syl>
                                        </verse>
                                    </note>
                                    <note xml:id="note-0000000489241935" dur.ppq="1" dur="4" oct="4" pname="e" stem.dir="up">
                                        <verse xml:id="verse-0000002121800829" n="1">
                                            <syl xml:id="syl-0000002140574568" con="d" wordpos="i">mea</syl>
                                        </verse>
                                    </note>
                                    <note xml:id="note-0000000064926164" dur.ppq="1" dur="4" oct="4" pname="e" stem.dir="up">
                                        <verse xml:id="verse-0000000292345672" n="1">
                                            <syl xml:id="syl-0000000011124968" con="s" wordpos="t">sure</syl>
                                        </verse>
                                    </note>
                                </layer>
                            </staff>
                        </measure>
                        <measure xml:id="measure-0000001468628641" right="end" n="3">
                            <staff xml:id="staff-0000000730162064" n="1">
                                <layer xml:id="layer-0000001112250690" n="1">
                                    <note xml:id="note-0000001899683342" dur.ppq="4" dur="1" oct="5" pname="c">
                                        <verse xml:id="verse-0000001338549045" n="1">
                                            <syl xml:id="syl-0000002102596990" con="s">3</syl>
                                        </verse>
                                    </note>
                                </layer>
                            </staff>
                        </measure>
                    </section>
                </score>
            </mdiv>
        </body>
    </music>
</mei>
```

**Comment by earboxer** [comment]
> I do not think having empty <mNum> is a good way to avoid the display of measure numbers.

@lpugin https://github.com/rism-ch/verovio/issues/1153#issuecomment-540940279

Although I think for the case of representing `measure[implicit=yes]`, it would still be fine. (otherwise I can't think of a way for `<measure number="X5" implicit="yes">` to be converted to MEI).

So I think the goal of this PR is still relevant.

I mentioned

> MusicXML can't represent explicit measure numbers

but they have the `text` attribute for measures, so you could do

```xml
<measure number="2"></measure>
<measure number="X2" text="2.5"></measure>
<measure number="3"></measure>
<measure number="X3" implicit="yes"></measure>
<measure number="4"></measure>
```

> The implicit attribute of the measure element should be set to "yes" rather than setting the text attribute to an empty string.

If we support measure text attribute, then that *could* be used as explicit, making MEI like

```xml
<measure n="2"></measure>
<measure n="X2">
  <mNum>2.5</mNum>
</measure>
<measure n="3"></measure>
<measure n="X3">
  <mNum/>
</measure>
<measure n="4"></measure>
```

Though I haven't seen any examples of the measure text attribute being used in MusicXML; I don't know if it's expected that it would make the measure's number explicit.

**Comment by lpugin** [comment]
@earboxer there are some issues with the method added to Object. Do we still need this?

**Comment by lpugin** [comment]
> What are the concerns with Object *Object::GetChild(int idx, const ClassId classId)?

As it is now it does not work because it simply looks for the first child of a classId, and then picks then nth next sibling. What you should do is use `Object::FindAllChildByComparison` with the appropriate parameters, and then return the object at the idx position of the filled vector (if any). Does this makes sense?

**Comment by earboxer** [comment]
> As it is now it does not work because it simply looks for the first child of a classId, and then picks then nth next sibling. What you should do is use Object::FindAllChildByComparison with the appropriate parameters, and then return the object at the idx position of the filled vector (if any). Does this makes sense?

Yes, I see it now.

> A simple approach would be to only allow integers for measure@n, or suppress the conversion to @n when implicit="yes" respectively. When there is no value Verovio wouldn't try to print a measure number.

I tried this first (not running `measure->SetN(measureNum);`). It resulted in unmatched ties IIRC.

**Comment by earboxer** [comment]
@lpugin I updated it to fix the issue.

File: src/object.cpp [hunk 1]
```diff
@@ -544,6 +544,17 @@ Object *Object::GetChild(int idx) const
     return m_children.at(idx);
 }
 
+Object *Object::GetChild(int idx, const ClassId classId)
+{
+    ArrayOfObjects objects;
+    ClassIdComparison matchClassId(classId);
+    this->FindAllChildByComparison(&objects, &matchClassId);
```
Comment (lpugin) [inline src/object.cpp]: I think we want to add a deepness parameter of 1, don't we?

Comment (earboxer) [reply src/object.cpp]: Thanks! 797169a6f876eaf772e885e3c5d03b0e18d4d5fc

**Comment by earboxer** [comment]
@lpugin This should be good now.

**Pull request merged**